Rename the custom post type and block namespace

kienstra opened the pull request:
#### Changes
* Changes the slug of the `block_lab` CPT
* Changes the block namespace in the CPT `post_content`. It's now `block-lab/`

#### Testing instructions
* Run PHPUnit tests
* There's no way to functionally test this yet. It isn't connected to anything now, as the UI is not yet decided

kienstra committed:
Begin migration with the CPT, leaving the new name a placeholder

Thanks to Weston for inspiration in migrate_single()
https://github.com/ampproject/amp-wp/blob/4880f0f58daaf07685854be8574ff25d76ff583e/includes/validation/class-amp-validated-url-post-type.php#L165-L170
The CPT needs a new post_type
and the block namespace in the post_content.

kienstra committed:
Correct the PHP DocBlock of migrate_all()

Remove a typo,
and clarify that this CTP is for a block config.

kienstra commented:
# Request for review

Hi @lukecarbis,
When you have a chance, could you please review this?

This is just migrating the `block_lab` CPT. There's no way to actually use this yet, as there isn't a UI yet.

So this shouldn't be merged yet, maybe it'll remain a feature branch for the migration work.

This will probably be changed slightly as the UI develops. Maybe there'll be error messages, or some indication of completion.

lukecarbis commented on php/blocks/migration/class-post-type.php:
```diff
@@ -0,0 +1,146 @@
+/**
+ * Class Post_Type
+ */
+final class Post_Type {
```
We haven't used `final` anywhere else. in Block Lab. Happy to include it – I can't imagine this class ever needing a child, but just wanted to call that out for consistency.

kienstra replied:
Thanks, good point. Yeah, it's probably not needed. We were using it with the AMP plugin so classes couldn't be extended, but I don't think people are going to be extending this.

lukecarbis commented on php/blocks/migration/class-post-type.php:
```diff
@@ -0,0 +1,146 @@
+
+		while ( ! empty( $posts ) ) {
+			foreach ( $posts as $post ) {
+				$this->migrate_single( $post );
```
Should we break and return a `WP_Error` or something like that if this returns `false`?

kienstra replied:
Good question, that might depend on the error feedback in the UI.

If there are 25 `block_lab` posts, and only the 10th has an error in migration, should we keep migrating the other 15, or stop the whole migration and display an error in the UI?

I think this `block_lab` post migration is less prone to error than the normal `post` or `page` migration, but we won't know until it's used.

lukecarbis replied:
That's a really good question. Let's leave this as it is now, so that it can continue if it hits an error, but keep that in mind as we develop the UI.

kienstra replied:
Thanks, sounds good.

kienstra replied:
If the migration of all of the `block_lab` posts failed, we'd definitely want some kind of feedback.

lukecarbis reviewed (approved):
Looks good! Couple of small questions, but happy for this to be merged as is.

kienstra committed:
Remove unnecessary  final before class

This is probably pedantic,
as there probably won't be anybody
extending this class.

kienstra committed:
Correct an issue when migrating many posts, add more tests

When attempting to migrate over 100 posts,
it looks like it doesn't migrate them.
So remove the 'offset' parameter,
which skipped over posts.

kienstra commented:
Hi @lukecarbis, 
Thanks for reviewing this.

If it's alright, I pushed [another commit](https://github.com/getblocklab/block-lab/pull/588/commits/f63f28b1f930122421bafd4e814505389623e969) that fixes an issue and adds more tests.

Also, the target branch is now `add/full-migration`, which is branched off `develop`.

Maybe that could be the feature branch for all of the migration work. It seems a little strange merging this into `develop` now, as there's no way to use the code yet.

lukecarbis commented on php/blocks/migration/class-post-type.php:
```diff
@@ -143,4 +119,20 @@ public function migrate_single( WP_Post $post ) {
+		$query = new WP_Query(
+			[
+				'post_type'      => $this->previous_post_type_slug,
+				'posts_per_page' => 10,
```
I don't understand how this can work without an offset parameter?

kienstra replied:
Yeah, this is confusing.

When the posts are migrated, they have their `post_type` changed. So this `WP_Query` won't find posts that were migrated, as it's looking for another `post_type`.

When this did have an `offset`, the offset eventually got high enough that there were no posts.

lukecarbis replied:
Ahhh – that makes total sense. Maybe leave a comment in there to that effect?

kienstra replied:
Sure, good idea. I missed that in the first iteration of this 😄

kienstra replied:
https://github.com/getblocklab/block-lab/pull/588/commits/3a3afefc054316cb48c20977b6b1c2f99cae1ba9 adds a comment about that

lukecarbis commented on php/blocks/migration/class-post-type.php:
```diff
@@ -132,7 +108,7 @@ public function migrate_single( WP_Post $post ) {
 			$wpdb->posts,
 			[
-				'post_type'    => $this->new_post_type_slug,
+				'post_type'    => sanitize_key( $this->new_post_type_slug ),
```
Good call.

kienstra replied:
Thanks

lukecarbis commented on tests/php/unit/blocks/migration/test-class-post-type.php:
```diff
@@ -147,6 +148,97 @@ public function test_migrate_all() {
+		);
+
+		// The block_lab CPT posts should now have the new post_type and migrated post_content.
+		$this->assertEquals( $number_of_block_lab_posts, $queried_new_cpt_posts->post_count );
```
How does this test ensure that the same set of posts weren't just looped through during migration?

e.g.
initial state: `a`,`b`,`c`,`d`,`e`,`f`,`g`,`h`,`i`,`j`,`k`,`l`,`m`,`n`,`o`,`p`
end state: `a`,`b`,`c`,`a`,`b`,`c`,`a`,`b`,`c`,`a`,`b`,`c`,`a`,`b`,`c`,`a`

(both sets have a count of 16)

lukecarbis replied:
Feel free to ignore this if you think it's safe enough as is.

kienstra replied:
Hm, this tests that the right amount of posts had their `post_type` changed, as they'll have the `self::NEW_POST_TYPE_SLUG`:

https://github.com/getblocklab/block-lab/blob/f63f28b1f930122421bafd4e814505389623e969/tests/php/unit/blocks/migration/test-class-post-type.php#L184-L189

Still, that `assertEquals()`  doesn't ensure that the `post_content` was changed.

lukecarbis commented:
@kienstra Great idea merging into `add/full-migration`. 👍

kienstra committed:
Add a comment about the lack of 'offset' parameter

It shouldn't be needed,
as the migration changes the post_type.

kienstra commented:
Also, we should be able to do automated e2e tests of the migration, as long as the new plugin is in a public repo that Travis can access.

It should be able to verify that the migration of a post worked and the new plugin's block looks right.
